File: python/test/test_attributes.py
#!/usr/bin/env python2

import unittest
from krpc import _Attributes as Attributes

class TestTypes(unittest.TestCase):

    def test_is_a_procedure(self):
        self.assertTrue(Attributes.is_a_procedure([]))
        self.assertFalse(Attributes.is_a_procedure(['Property.Get(PropertyName)']))
        self.assertFalse(Attributes.is_a_procedure(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_procedure(['Class.Method(ClassName,MethodName)']))
        self.assertFalse(Attributes.is_a_procedure(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertFalse(Attributes.is_a_procedure(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_property_accessor(self):
        self.assertFalse(Attributes.is_a_property_accessor([]))
        self.assertTrue(Attributes.is_a_property_accessor(['Property.Get(PropertyName)']))
        self.assertTrue(Attributes.is_a_property_accessor(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_property_accessor(['Class.Method(ClassName,MethodName)']))
        self.assertFalse(Attributes.is_a_property_accessor(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertFalse(Attributes.is_a_property_accessor(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_property_getter(self):
        self.assertFalse(Attributes.is_a_property_getter([]))
        self.assertTrue(Attributes.is_a_property_getter(['Property.Get(PropertyName)']))
        self.assertFalse(Attributes.is_a_property_getter(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_property_getter(['Class.Method(ClassName,MethodName)']))
        self.assertFalse(Attributes.is_a_property_getter(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertFalse(Attributes.is_a_property_getter(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_property_setter(self):
        self.assertFalse(Attributes.is_a_property_setter([]))
        self.assertFalse(Attributes.is_a_property_setter(['Property.Get(PropertyName)']))
        self.assertTrue(Attributes.is_a_property_setter(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_property_setter(['Class.Method(ClassName,MethodName)']))
        self.assertFalse(Attributes.is_a_property_setter(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertFalse(Attributes.is_a_property_setter(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_class_method(self):
        self.assertFalse(Attributes.is_a_class_method([]))
        self.assertFalse(Attributes.is_a_class_method(['Property.Get(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_method(['Property.Set(PropertyName)']))
        self.assertTrue(Attributes.is_a_class_method(['Class.Method(ClassName,MethodName)']))
        self.assertFalse(Attributes.is_a_class_method(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertFalse(Attributes.is_a_class_method(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_class_property_accessor(self):
        self.assertFalse(Attributes.is_a_class_property_accessor([]))
        self.assertFalse(Attributes.is_a_class_property_accessor(['Property.Get(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_accessor(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_accessor(['Class.Method(ClassName,MethodName)']))
        self.assertTrue(Attributes.is_a_class_property_accessor(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertTrue(Attributes.is_a_class_property_accessor(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_class_property_getter(self):
        self.assertFalse(Attributes.is_a_class_property_getter([]))
        self.assertFalse(Attributes.is_a_class_property_getter(['Property.Get(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_getter(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_getter(['Class.Method(ClassName,MethodName)']))
        self.assertTrue(Attributes.is_a_class_property_getter(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_getter(['Class.Property.Set(ClassName,PropertyName)']))

    def test_is_a_class_property_setter(self):
        self.assertFalse(Attributes.is_a_class_property_setter([]))
        self.assertFalse(Attributes.is_a_class_property_setter(['Property.Get(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_setter(['Property.Set(PropertyName)']))
        self.assertFalse(Attributes.is_a_class_property_setter(['Class.Method(ClassName,MethodName)']))
        self.assertFalse(Attributes.is_a_class_property_setter(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertTrue(Attributes.is_a_class_property_setter(['Class.Property.Set(ClassName,PropertyName)']))

    def test_get_property_name(self):
        self.assertRaises(ValueError, Attributes.get_property_name, [])
        self.assertEqual('PropertyName', Attributes.get_property_name(['Property.Get(PropertyName)']))
        self.assertEqual('PropertyName', Attributes.get_property_name(['Property.Set(PropertyName)']))
        self.assertRaises(ValueError, Attributes.get_property_name, ['Class.Method(ClassName,MethodName)'])
        self.assertRaises(ValueError, Attributes.get_property_name, ['Class.Property.Get(ClassName,PropertyName)'])
        self.assertRaises(ValueError, Attributes.get_property_name, ['Class.Property.Set(ClassName,PropertyName)'])

    def test_get_class_name(self):
        self.assertRaises(ValueError, Attributes.get_class_name, [])
        self.assertRaises(ValueError, Attributes.get_class_name, ['Property.Get(PropertyName)'])
        self.assertRaises(ValueError, Attributes.get_class_name, ['Property.Set(PropertyName)'])
        self.assertEqual('ClassName', Attributes.get_class_name(['Class.Method(ClassName,MethodName)']))
        self.assertEqual('ClassName', Attributes.get_class_name(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertEqual('ClassName', Attributes.get_class_name(['Class.Property.Set(ClassName,PropertyName)']))

    def test_get_class_method_name(self):
        self.assertRaises(ValueError, Attributes.get_class_method_name, [])
        self.assertRaises(ValueError, Attributes.get_class_method_name, ['Property.Get(PropertyName)'])
        self.assertRaises(ValueError, Attributes.get_class_method_name, ['Property.Set(PropertyName)'])
        self.assertEqual('MethodName', Attributes.get_class_method_name(['Class.Method(ClassName,MethodName)']))
        self.assertRaises(ValueError, Attributes.get_class_method_name, ['Class.Property.Get(ClassName,PropertyName)'])
        self.assertRaises(ValueError, Attributes.get_class_method_name, ['Class.Property.Set(ClassName,PropertyName)'])

    def test_get_class_property_name(self):
        self.assertRaises(ValueError, Attributes.get_class_property_name, [])
        self.assertRaises(ValueError, Attributes.get_class_property_name, ['Property.Get(PropertyName)'])
        self.assertRaises(ValueError, Attributes.get_class_property_name, ['Property.Set(PropertyName)'])
        self.assertRaises(ValueError, Attributes.get_class_property_name, ['Class.Method(ClassName,MethodName)'])
        self.assertEqual('PropertyName', Attributes.get_class_property_name(['Class.Property.Get(ClassName,PropertyName)']))
        self.assertEqual('PropertyName', Attributes.get_class_property_name(['Class.Property.Set(ClassName,PropertyName)']))

    def test_get_return_type_attributes(self):
        self.assertEqual([], Attributes.get_return_type_attrs([]))
        self.assertEqual([], Attributes.get_return_type_attrs(['Class.Method(ClassName,MethodName)']))
        self.assertEqual(['Class(ClassName)'], Attributes.get_return_type_attrs(['ReturnType.Class(ClassName)']))
        self.assertEqual(['Class(ClassName)'], Attributes.get_return_type_attrs(['Class.Method(ClassName,MethodName)', 'ReturnType.Class(ClassName)']))

    def test_get_parameter_type_attributes(self):
        self.assertEqual([], Attributes.get_parameter_type_attrs(0, []))
        self.assertEqual([], Attributes.get_parameter_type_attrs(0, ['Class.Method(ClassName,MethodName)']))
        self.assertEqual([], Attributes.get_parameter_type_attrs(0, ['ReturnType.Class(ClassName)']))
        self.assertEqual([], Attributes.get_parameter_type_attrs(0, ['Class.Method(ClassName,MethodName)', 'ReturnType.Class(ClassName)']))
        self.assertEqual([], Attributes.get_parameter_type_attrs(1, ['ParameterType(2).Class(ClassName)']))
        self.assertEqual(['Class(ClassName)'], Attributes.get_parameter_type_attrs(2, ['ParameterType(2).Class(ClassName)']))
        self.assertEqual(['Class(ClassName2)'], Attributes.get_parameter_type_attrs(2, ['ParameterType(0).Class(ClassName1)', 'ParameterType(2).Class(ClassName2)']))
        self.assertEqual(['Class(ClassName)'], Attributes.get_parameter_type_attrs(1, ['Class.Method(ClassName,MethodName)', 'ParameterType(1).Class(ClassName)']))

if __name__ == '__main__':
    unittest.main()

File: lua/krpc/test/test_attributes.lua
local luaunit = require 'luaunit'
local class = require 'pl.class'
local List = require 'pl.List'
local Attributes = require 'krpc.attributes'

local TestAttributes = class()

function TestAttributes:test_is_a_procedure()
  luaunit.assertTrue(Attributes.is_a_procedure(List{}))
  luaunit.assertFalse(Attributes.is_a_procedure(List{'Property.Get(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_procedure(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_procedure(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_procedure(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_procedure(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_procedure(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_property_accessor()
  luaunit.assertFalse(Attributes.is_a_property_accessor(List{}))
  luaunit.assertTrue(Attributes.is_a_property_accessor(List{'Property.Get(PropertyName)'}))
  luaunit.assertTrue(Attributes.is_a_property_accessor(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_accessor(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_property_accessor(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_property_accessor(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_accessor(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_property_getter()
  luaunit.assertFalse(Attributes.is_a_property_getter(List{}))
  luaunit.assertTrue(Attributes.is_a_property_getter(List{'Property.Get(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_getter(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_getter(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_property_getter(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_property_getter(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_getter(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_property_setter()
  luaunit.assertFalse(Attributes.is_a_property_setter(List{}))
  luaunit.assertFalse(Attributes.is_a_property_setter(List{'Property.Get(PropertyName)'}))
  luaunit.assertTrue(Attributes.is_a_property_setter(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_setter(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_property_setter(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_property_setter(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_property_setter(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_class_method()
  luaunit.assertFalse(Attributes.is_a_class_method(List{}))
  luaunit.assertFalse(Attributes.is_a_class_method(List{'Property.Get(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_method(List{'Property.Set(PropertyName)'}))
  luaunit.assertTrue(Attributes.is_a_class_method(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_class_method(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_class_method(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_method(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_class_property_accessor()
  luaunit.assertFalse(Attributes.is_a_class_property_accessor(List{}))
  luaunit.assertFalse(Attributes.is_a_class_property_accessor(List{'Property.Get(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_accessor(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_accessor(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_accessor(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertTrue(Attributes.is_a_class_property_accessor(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertTrue(Attributes.is_a_class_property_accessor(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_class_property_getter()
  luaunit.assertFalse(Attributes.is_a_class_property_getter(List{}))
  luaunit.assertFalse(Attributes.is_a_class_property_getter(List{'Property.Get(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_getter(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_getter(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_getter(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertTrue(Attributes.is_a_class_property_getter(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_getter(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_is_a_class_property_setter()
  luaunit.assertFalse(Attributes.is_a_class_property_setter(List{}))
  luaunit.assertFalse(Attributes.is_a_class_property_setter(List{'Property.Get(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_setter(List{'Property.Set(PropertyName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_setter(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_setter(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertFalse(Attributes.is_a_class_property_setter(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertTrue(Attributes.is_a_class_property_setter(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_get_property_name()
  luaunit.assertError(Attributes.get_property_name, List{})
  luaunit.assertEquals('PropertyName', Attributes.get_property_name(List{'Property.Get(PropertyName)'}))
  luaunit.assertEquals('PropertyName', Attributes.get_property_name(List{'Property.Set(PropertyName)'}))
  luaunit.assertError(ValueError, Attributes.get_property_name, List{'Class.Method(ServiceName.ClassName,MethodName)'})
  luaunit.assertError(Attributes.get_property_name, List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'})
  luaunit.assertError(Attributes.get_property_name, List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'})
  luaunit.assertError(Attributes.get_property_name, List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'})
end

function TestAttributes:test_get_service_name()
  luaunit.assertError(Attributes.get_service_name, List{})
  luaunit.assertError(Attributes.get_service_name, List{'Property.Get(PropertyName)'})
  luaunit.assertError(Attributes.get_service_name, List{'Property.Set(PropertyName)'})
  luaunit.assertEquals('ServiceName', Attributes.get_service_name(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals('ServiceName', Attributes.get_service_name(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals('ServiceName', Attributes.get_service_name(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertEquals('ServiceName', Attributes.get_service_name(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_get_class_name()
  luaunit.assertError(Attributes.get_class_name, List{})
  luaunit.assertError(Attributes.get_class_name, List{'Property.Get(PropertyName)'})
  luaunit.assertError(Attributes.get_class_name, List{'Property.Set(PropertyName)'})
  luaunit.assertEquals('ClassName', Attributes.get_class_name(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals('ClassName', Attributes.get_class_name(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals('ClassName', Attributes.get_class_name(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertEquals('ClassName', Attributes.get_class_name(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_get_class_method_name()
  luaunit.assertError(Attributes.get_class_method_name, List{})
  luaunit.assertError(Attributes.get_class_method_name, List{'Property.Get(PropertyName)'})
  luaunit.assertError(Attributes.get_class_method_name, List{'Property.Set(PropertyName)'})
  luaunit.assertEquals('MethodName', Attributes.get_class_method_name(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals('MethodName', Attributes.get_class_method_name(List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'}))
  luaunit.assertError(Attributes.get_class_method_name, List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'})
  luaunit.assertError(Attributes.get_class_method_name, List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'})
end

function TestAttributes:test_get_class_property_name()
  luaunit.assertError(Attributes.get_class_property_name, List{})
  luaunit.assertError(Attributes.get_class_property_name, List{'Property.Get(PropertyName)'})
  luaunit.assertError(Attributes.get_class_property_name, List{'Property.Set(PropertyName)'})
  luaunit.assertError(Attributes.get_class_property_name, List{'Class.Method(ServiceName.ClassName,MethodName)'})
  luaunit.assertError(Attributes.get_class_property_name, List{'Class.StaticMethod(ServiceName.ClassName,MethodName)'})
  luaunit.assertEquals('PropertyName', Attributes.get_class_property_name(List{'Class.Property.Get(ServiceName.ClassName,PropertyName)'}))
  luaunit.assertEquals('PropertyName', Attributes.get_class_property_name(List{'Class.Property.Set(ServiceName.ClassName,PropertyName)'}))
end

function TestAttributes:test_get_return_type_attributes()
  luaunit.assertEquals(List{}, Attributes.get_return_type_attrs(List{}))
  luaunit.assertEquals(List{}, Attributes.get_return_type_attrs(List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals(List{'Class(ServiceName.ClassName)'}, Attributes.get_return_type_attrs(List{'ReturnType.Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{'Class(ServiceName.ClassName)'}, Attributes.get_return_type_attrs(List{'Class.Method(ServiceName.ClassName,MethodName)', 'ReturnType.Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{'List(string)'}, Attributes.get_return_type_attrs(List{'ReturnType.List(string)'}))
  luaunit.assertEquals(List{'Dictionary(int32,string)'}, Attributes.get_return_type_attrs(List{'ReturnType.Dictionary(int32,string)'}))
  luaunit.assertEquals(List{'Set(string)'}, Attributes.get_return_type_attrs(List{'ReturnType.Set(string)'}))
  luaunit.assertEquals(List{'List(Dictionary(int32,string))'}, Attributes.get_return_type_attrs(List{'ReturnType.List(Dictionary(int32,string))'}))
  luaunit.assertEquals(List{'Dictionary(int32,List(ServiceName.ClassName))'}, Attributes.get_return_type_attrs(List{'ReturnType.Dictionary(int32,List(ServiceName.ClassName))'}))
end

function TestAttributes:test_get_parameter_type_attributes()
  luaunit.assertEquals(List{}, Attributes.get_parameter_type_attrs(1, List{}))
  luaunit.assertEquals(List{}, Attributes.get_parameter_type_attrs(1, List{'Class.Method(ServiceName.ClassName,MethodName)'}))
  luaunit.assertEquals(List{}, Attributes.get_parameter_type_attrs(1, List{'ReturnType.Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{}, Attributes.get_parameter_type_attrs(1, List{'Class.Method(ServiceName.ClassName,MethodName)', 'ReturnType.Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{}, Attributes.get_parameter_type_attrs(2, List{'ParameterType(2).Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{'Class(ServiceName.ClassName)'}, Attributes.get_parameter_type_attrs(3, List{'ParameterType(2).Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{'Class(ServiceName.ClassName2)'}, Attributes.get_parameter_type_attrs(3, List{'ParameterType(0).Class(ServiceName.ClassName1)', 'ParameterType(2).Class(ServiceName.ClassName2)'}))
  luaunit.assertEquals(List{'Class(ServiceName.ClassName)'}, Attributes.get_parameter_type_attrs(2, List{'Class.Method(ServiceName.ClassName,MethodName)', 'ParameterType(1).Class(ServiceName.ClassName)'}))
  luaunit.assertEquals(List{'List(string)'}, Attributes.get_parameter_type_attrs(2, List{'ParameterType(1).List(string)'}))
  luaunit.assertEquals(List{'Dictionary(int32,string)'}, Attributes.get_parameter_type_attrs(2, List{'ParameterType(1).Dictionary(int32,string)'}))
  luaunit.assertEquals(List{'Set(string)'}, Attributes.get_parameter_type_attrs(2, List{'ParameterType(1).Set(string)'}))
  luaunit.assertEquals(List{'List(Dictionary(int32,string))'}, Attributes.get_parameter_type_attrs(2, List{'ParameterType(1).List(Dictionary(int32,string))'}))
  luaunit.assertEquals(List{'Dictionary(int32,List(ServiceName.ClassName))'}, Attributes.get_parameter_type_attrs(2, List{'ParameterType(1).Dictionary(int32,List(ServiceName.ClassName))'}))
end

return TestAttributes
